### PR TITLE
perf: improve performance of isomorphicEncode

### PR DESCRIPTION
**Pull request by tsctx**
<details>

<summary>Benchmark Script</summary>

```js
import { bench, group, run } from 'mitata'
import { isomorphicEncode } from '../../lib/web/fetch/util.js'

const characters =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
const charactersLength = characters.length

function generateAsciiString (length) {
  let result = ''
  for (let i = 0; i < length; ++i) {
    result += characters[Math.floor(Math.random() * charactersLength)]
  }
  return result
}

const invalidIsomorphicEncodeValueRegex = /[^\x00-\xFF]/ // eslint-disable-line

function isomorphicEncode1 (input) {
  for (let i = 0; i < input.length; i++) {
    if (input.charCodeAt(i) > 0xff) {
      throw new TypeError('Unreachable')
    }
  }
  return input
}

function isomorphicEncode2 (input) {
  if (invalidIsomorphicEncodeValueRegex.test(input)) {
    throw new TypeError('Unreachable')
  }
  return input
}

const settings = {
  small: 10,
  middle: 30,
  long: 70
}

for (const [runName, length] of Object.entries(settings)) {
  const value = generateAsciiString(length);

  [
    { name: `${runName} (valid)`, value },
    {
      name: `${runName} (invalid)`,
      value: `${value.slice(0, -1)}${String.fromCharCode(0xff + 1)}`
    }
  ].forEach(({ name, value }) => {
    group(name, () => {
      [
        {
          name: 'original',
          fn: isomorphicEncode
        },
        {
          name: 'String#charCodeAt',
          fn: isomorphicEncode1
        },
        {
          name: 'RegExp#test',
          fn: isomorphicEncode2
        }
      ].forEach(({ name, fn }) => {
        bench(name, () => {
          try {
            return fn(value)
          } catch (err) {}
        })
      })
    })
  })
}

await run()
```

</details>

```
benchmark              time (avg)             (min … max)       p75       p99      p999
--------------------------------------------------------- -----------------------------
• small (valid)
--------------------------------------------------------- -----------------------------
original            56.48 ns/iter     (46.19 ns … 973 ns)  55.57 ns    122 ns    268 ns
String#charCodeAt   44.82 ns/iter     (33.79 ns … 875 ns)  57.47 ns    123 ns    324 ns
RegExp#test         39.23 ns/iter   (23.24 ns … 2'271 ns)  47.17 ns  88.53 ns    483 ns

summary for small (valid)
  RegExp#test
   1.14x faster than String#charCodeAt
   1.44x faster than original

• small (invalid)
--------------------------------------------------------- -----------------------------
original              162 µs/iter  (76'200 ns … 4'403 µs)    171 µs    425 µs  2'893 µs
String#charCodeAt  20'686 ns/iter  (11'300 ns … 5'352 µs) 25'200 ns 39'300 ns    204 µs
RegExp#test        15'999 ns/iter (10'136 ns … 35'269 ns) 20'371 ns 32'735 ns 35'269 ns

summary for small (invalid)
  RegExp#test
   1.29x faster than String#charCodeAt
   10.14x faster than original

• middle (valid)
--------------------------------------------------------- -----------------------------   
original              319 ns/iter     (160 ns … 7'139 ns)    334 ns  1'397 ns  7'139 ns
String#charCodeAt     202 ns/iter     (113 ns … 2'750 ns)    233 ns    463 ns  1'956 ns
RegExp#test         33.93 ns/iter   (23.19 ns … 2'237 ns)  34.57 ns    120 ns    285 ns

summary for middle (valid)
  RegExp#test
   5.96x faster than String#charCodeAt
   9.4x faster than original

• middle (invalid)
--------------------------------------------------------- -----------------------------   
original              116 µs/iter  (88'000 ns … 1'768 µs)    107 µs    274 µs    661 µs
String#charCodeAt  29'194 ns/iter (15'500 ns … 12'211 µs) 32'200 ns 55'800 ns    381 µs
RegExp#test        29'225 ns/iter  (16'100 ns … 3'467 µs) 33'000 ns 53'600 ns    221 µs

summary for middle (invalid)
  String#charCodeAt
   1x faster than RegExp#test
   3.96x faster than original

• long (valid)
--------------------------------------------------------- -----------------------------   
original              657 ns/iter     (352 ns … 3'251 ns)    748 ns  1'874 ns  3'251 ns
String#charCodeAt     307 ns/iter     (225 ns … 1'323 ns)    308 ns    630 ns  1'323 ns
RegExp#test         28.94 ns/iter     (23.63 ns … 230 ns)  25.63 ns    102 ns    156 ns

summary for long (valid)
  RegExp#test
   10.61x faster than String#charCodeAt
   22.72x faster than original

• long (invalid)
--------------------------------------------------------- -----------------------------   
original              155 µs/iter  (90'600 ns … 6'604 µs)    182 µs    338 µs  2'145 µs
String#charCodeAt  29'024 ns/iter  (16'500 ns … 3'502 µs) 33'500 ns 49'000 ns    212 µs
RegExp#test        25'481 ns/iter  (16'800 ns … 3'543 µs) 30'300 ns 47'300 ns    297 µs

summary for long (invalid)
  RegExp#test
   1.14x faster than String#charCodeAt
   6.09x faster than original
```

### CODE REVIEW

File: benchmarks/fetch/isomorphic-encode.mjs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  // 2. Return a byte sequence whose length is equal to input’s code
+  //    point length and whose bytes have the same values as the
+  //    values of input’s code points, in the same order
+  return input
+}
+
+/**
+ * @see https://infra.spec.whatwg.org/#isomorphic-encode
+ * @param {string} input
+ */
```
Comment (tsctx) [inline benchmarks/fetch/isomorphic-encode.mjs:34]:
```suggestion
```

File: benchmarks/fetch/isomorphic-encode.mjs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  //    point length and whose bytes have the same values as the
+  //    values of input’s code points, in the same order
+  return input
+}
+
+const settings = {
+  small: `${generateAsciiString(10)}`,
+  middle: `${generateAsciiString(30)}`,
+  long: `${generateAsciiString(70)}`
+}
```
Comment (tsctx) [inline benchmarks/fetch/isomorphic-encode.mjs:50]:
```suggestion
const settings = {
  small: 10,
  middle: 30,
  long: 70
}
```

File: benchmarks/fetch/isomorphic-encode.mjs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+for (const [runName, value] of Object.entries(settings)) {
+  [
```
Comment (tsctx) [inline benchmarks/fetch/isomorphic-encode.mjs:53]:
```suggestion
for (const [runName, length] of Object.entries(settings)) {
  const value = generateAsciiString(length)
  [
```

**Review by ronag — APPROVED** [review]

**Comment by Uzlopak** [comment]
interesting. I was also eyeing it, when I created #3091 . But one thing: I would maybe not write "Unreachable" as a  message. Maybe if we have AssertionError, then use that. But yeah.

**Comment by Uzlopak** [comment]
Can you please not remove the benchmark?

**Comment by Uzlopak** [comment]
Can you add a unit test for the error case? If we touch it we can immediatly improve the test coverage ;)

**Comment by KhafraDev** [comment]
why is the error case unreachable?

File: lib/web/fetch/util.js [hunk 4]
```diff
@@ -1098,14 +1098,16 @@ function readableStreamClose (controller) {
...
  * @see https://infra.spec.whatwg.org/#isomorphic-encode
  * @param {string} input
  */
 function isomorphicEncode (input) {
   // 1. Assert: input contains no code points greater than U+00FF.
-  for (let i = 0; i < input.length; i++) {
-    assert(input.charCodeAt(i) <= 0xFF)
+  if (invalidIsomorphicEncodeValueRegex.test(input)) {
+    throw new TypeError('Unreachable')
   }
```
Comment (tsctx) [inline lib/web/fetch/util.js:1111]:
```suggestion
    assert(!invalidIsomorphicEncodeValueRegex.test(input))
```

**Review by KhafraDev — APPROVED** [review]

File: lib/web/fetch/util.js [hunk 5]
```diff
@@ -1098,15 +1098,15 @@ function readableStreamClose (controller) {
...
 /**
  * @see https://infra.spec.whatwg.org/#isomorphic-encode
  * @param {string} input
  */
 function isomorphicEncode (input) {
   // 1. Assert: input contains no code points greater than U+00FF.
-  for (let i = 0; i < input.length; i++) {
-    assert(input.charCodeAt(i) <= 0xFF)
-  }
+    assert(!invalidIsomorphicEncodeValueRegex.test(input))
```
Comment (tsctx) [inline lib/web/fetch/util.js:1109]:
```suggestion
  assert(!invalidIsomorphicEncodeValueRegex.test(input))
```

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**